[3.9] Pre-Update Checker

SamuelSchepp opened the pull request:
### Summary of Changes

Based on https://github.com/joomla/joomla-cms/pull/16494 we continued developing the Pre-Update Checker for Joomla! 3.9. Like the Joomla! installation process, the com_joomlaupdate component now checks PHP options, settings and extension compatibility.
The evaluation process is currently a copy of the installation process pre-checker.
The translation of extension names is duplicated from com_installer.
Future PRs should address these issues by moving these functions to a core library.

#### Conditional checks:
##### If next J! version >= 4:
* Check for unsupported database types (sqlsrv, sqlazure)
##### If PHP version of next J! < 7:
* Magic Quotes (+ Runtime)
* Register Globals
* Safe Mode

### Testing Instructions

These instructions are based on a Joomla 3.9 installation.
Change values accordingly.
The path <code>/</code> represents the Joomla! installation directory.

#### Test Joomla! Pre-Update Check

1. Create file `list.xml` at `/`
```xml
<extensionset>
	<extension element="joomla" type="file" version="4.0.0" targetplatformversion="3.9" detailsurl="http://localhost/extension_sts.xml" />
</extensionset>
```

2. Create file `extension_sts.xml` at `/`
```xml
<updates>
	<update>
		<name>Joomla! 3.7</name>
		<description>Joomla! 3.7 CMS</description>
		<element>joomla</element>
		<type>file</type>
		<version>4.0.0</version>
		<infourl title="Joomla!">https://www.joomla.org/announcements/release-news/5711-joomla-3-7-5-release.html</infourl>
		<downloads>
			<downloadurl type="full" format="zip">https://downloads.joomla.org/cms/joomla3/3-7-5/Joomla_3.7.5-Stable-Update_Package.zip</downloadurl>
		</downloads>
		<tags>
			<tag>stable</tag>
		</tags>
		<maintainer>Joomla! PLT</maintainer>
		<maintainerurl>https://www.joomla.org</maintainerurl>
		<section>STS</section>
		<targetplatform name="joomla" version="3.[23456789]" />
		<php_minimum>7.0</php_minimum>
	</update>
</updates>
```

3. Components -> Joomla! Update -> Options

Custom URL: `localhost/list.xml`

4. Save & Close

There should be a fake update to Joomla 4.

5. Change your PHP or webserver settings to trigger a failed compatibility check.

`display_errors = On`

`file_uploads = Off`

`output_buffering = On`

6. Run `chmod -w configuration.php` to disable write on the configuration file.

7. Visit the update site and check the compatibilty results.

#### Test Extension Compatibility

1. Install some extensions, which are not compatible with your hacked Joomla! target version (4.0.0).

2. Change the XML Update-URL of any incompatible extension in the table `#__update_sites` in the column `location` to an own custom XML file.

3. Put the following contents in your XML update file. Replace version 1.5 with the actual installed version of your selected extension.

```xml
<updates>
	<update>
		<version>1.5</version>
		<downloads>
			<downloadurl type="full" format="zip">https://fakeurl</downloadurl>
		</downloads>
		<targetplatform name="joomla" version="4.0.0" />
	</update>
</updates>
```

4. Check the extensions compatiblity result.

5. Change the column `manifest_cache` in the table `#__extensions` of the hacked extension by changing the field `version` to an older one. This should trigger a compatiblilty, but based on a newer version.

#### Test Database compatibility

Add your current database type (e.g. `mysqli`) to `$unsupportedDatabaseTypes` in function `isDatabaseTypeSupported()` in file `./administrator/components/com_joomlaupdate/models/default.php`.

### Expected result

You should see a full report on compatibility of the update. PHP options, settings and extensions should be marked accordingly.
An incompatible database type should cause the message "Your host doesn't support the minimum requirements for Joomla".
<img width="1037" alt="preupdate1" src="https://user-images.githubusercontent.com/11752441/41082853-eb9eb212-6a2f-11e8-99dc-82840f6f4429.png">
<img width="830" alt="preupdate2" src="https://user-images.githubusercontent.com/11752441/41082858-ed6607c6-6a2f-11e8-965e-a765cda8bd2a.png">

### Actual result

No results, since this is a new feature.

### Documentation Changes Required

The following page should be updated with recent screenshots and descriptions.
https://docs.joomla.org/J3.x:Updating_from_an_existing_version

Developed @icampus, @SamuelSchepp, @Kev1n337, @DrDreave

SamuelSchepp committed:
added hacked xml files to fake an update process

DrDreave committed:
Add functions for pre install check

DrDreave committed:
Implement php options pre update check

SamuelSchepp committed:
Showed phpOptions phpSettings in view.

SamuelSchepp committed:
Added better labels

Kev1n337 committed:
Add phpsettings for pre update check

SamuelSchepp committed:
worked on layout and built preset for 3rd party extensions html

SamuelSchepp committed:
implemented JS & HTML variables

SamuelSchepp committed:
fixed missing noticed desciptions

Kev1n337 committed:
Add method to get all non core extensions

DrDreave committed:
Add first version of client-side extension compatibility check

Added javascript to pre update check that checks the compatibility of 3rd party extensions based on data attributes containing the current version and the update server url.

SamuelSchepp committed:
Gluecode to bring JS and model together

SamuelSchepp committed:
Added lang constants to view

SamuelSchepp committed:
fixed missing ;

SamuelSchepp committed:
fixed missing ". Added current version column

DrDreave committed:
Add error handling to extension compatibility check

SamuelSchepp committed:
Added Access-Control-Allow-Origin proxy

DrDreave committed:
Use proxy to check extension compatibility

Now using "proxy" checking extension compatibility to prevent CORS errors.

DrDreave committed:
Localize extension compatibility check strings

DrDreave committed:
Fix check for update server url

SamuelSchepp committed:
removed label border from recommended column

DrDreave committed:
Further error handling and clean up

DrDreave committed:
Fix missing update url error handling

SamuelSchepp committed:
added warning label

SamuelSchepp committed:
added descriptive labels on bottom

DrDreave committed:
Test extension compatibility via regex

SamuelSchepp committed:
fixed some safari stuff

SamuelSchepp committed:
Added todo

SamuelSchepp committed:
Deleted list.xml (Debug file)

SamuelSchepp committed:
deleted extension_sts.xml (Debug file)

SamuelSchepp committed:
changed fetchURL interface

SamuelSchepp committed:
changed HTML field data to ID instead of URL

DrDreave committed:
Move extension compatibility check to the server

DrDreave committed:
Adjust response processing to JResponseJson

Kev1n337 committed:
Fetch file and build json

Kev1n337 committed:
Merge branch '3.9-dev-installation' of https://github.com/SamuelSchepp/joomla-cms into 3.9-dev-installation

SamuelSchepp committed:
fixed json response

SamuelSchepp committed:
Added functionality, that shows the compatible version only if its not the current

DrDreave committed:
Add comments to javascript

Kev1n337 committed:
Validate compatibility and return json

SamuelSchepp committed:
changed a variable name

SamuelSchepp committed:
removed unecessary function, resorted functions, added todo to functions

SamuelSchepp committed:
Changed missing compatibility tag to gray. Fixed JS bake.

SamuelSchepp committed:
Fixed some issues

SamuelSchepp committed:
Some style stuff

SamuelSchepp committed:
some refactoring & phpcs fixes

SamuelSchepp committed:
Fixed PHP error "undefined constant"

SamuelSchepp committed:
JText für nonCoreExtension - tabelle komprimierter

Kev1n337 committed:
Translate extension names

Kev1n337 committed:
Cleanup, remove unnecessary loop

SamuelSchepp committed:
Added PHP version check

SamuelSchepp committed:
Merge branch '3.9-dev-installation' of github.com:SamuelSchepp/joomla-cms into 3.9-dev-installation

* '3.9-dev-installation' of github.com:SamuelSchepp/joomla-cms:
  Cleanup, remove unnecessary loop

DrDreave committed:
Remove frontend check if compatible version matches current version

Kev1n337 committed:
Adapt variable names to database columns

Kev1n337 committed:
Merge branch '3.9-dev-installation' of https://github.com/SamuelSchepp/joomla-cms into 3.9-dev-installation

SamuelSchepp committed:
Changed warning label on non-matching installed version. Added description for warning label.

SamuelSchepp committed:
fixed undefinied php minimum version

DrDreave committed:
Don't check submodules' compatibility

DrDreave committed:
Clean up localization

Now using joomla core strings for trivial translations (yes, no, on, off).

DrDreave committed:
Optimize template

- Cleaned up code
- Added tooltips for notices

DrDreave committed:
Clean up localization file

- Remove unused localizations
- Remove break tag in localization

Kev1n337 committed:
Use helper method to determine if extension is core extension

Kev1n337 committed:
Fix label not being set sometimes

Kev1n337 committed:
Update doc

Kev1n337 committed:
Add new lines add end of file

Kev1n337 committed:
Replace spaces with tabs and remove empty tag

roland-d commented:
@wilsonge Can you have a look?

roland-d commented:
@SamuelSchepp I think it will be good to include the image of how it should look like on screen to your post as well.

brianteeman commented:
Is there really any point in adding to the preupdate checker tests for things that would have prevented joomla installing in the first place. For example if Joomla 3.9 can not be installed if you have magic quotes enabled then there is no need to test for it here as it can only ever give one result

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -79,3 +79,44 @@ COM_JOOMLAUPDATE_VIEW_UPLOAD_CAPTIVE_INTRO_BODY="Make sure that the update file
+COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_DESC1="*Extensions marked with <span class='label label-important'>No</span> or <span class='label'>Missing Compatibility Tag</span> might break your website. Please consult with the developer before upgrading."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_DESC2="*Extensions marked with <span class='label'>Missing Compatibility Tag</span> indicate the developer did not a compatibility tag in the respective extension's XML."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_DESC3="*Extensions marked with <span class='label label-warning'>Yes (X.X.X)</span> might require an update."
```
Can we have more meaningful KEYS please

SamuelSchepp committed:
Changed some language keys, fixed phpcs issues

SamuelSchepp committed:
Merge branch '3.9-dev-installation' of github.com:SamuelSchepp/joomla-cms into 3.9-dev-installation

* '3.9-dev-installation' of github.com:SamuelSchepp/joomla-cms:
  Replace spaces with tabs and remove empty tag

roland-d commented:
> Is there really any point in adding to the preupdate checker tests for things that would have prevented joomla installing in the first place.

Besides the PHP version all settings are hardcoded so we can't know beforehand if they have changed or not. The only thing we can do is filter out settings that are no longer relevant due to the PHP version, such as the Magic Quotes and Register Globals.

The other settings may have been correct during installation but may have changed afterwards due to server changes or moving sites. So I think having these checks in place is just informative to the user.

The only thing I think can be left out is the configuration.php check because this file is not going to be changed during an update.

SamuelSchepp committed:
Removed checks of Magic Quotes, Register Globals and configuration.php writeable

Kev1n337 commented:
We just improved the language keys and removed the magic quotes, register globals and configuration.php checks as these aren't required for an update to future Joomla versions. We think that the other checks are still useful as they could have changed after the installation.

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1001,309 @@ public function removePackageFiles()
-		$query = $db->getQuery(true);
+		/*
+		 * Check the PHP Version. It is already checked in JUpdate.
+		 * An Joomla! Update which is not supported by current PHP
```
Really pedantic but this should be "A Joomla! Update which is not supported by the current PHP

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1001,309 @@ public function removePackageFiles()
+		}
+		else
+		{
+			// Attempt to detect their existence; even pure PHP implementation of them will trigger a positive response, though.
```
impementations

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -79,3 +79,44 @@ COM_JOOMLAUPDATE_VIEW_UPLOAD_CAPTIVE_INTRO_BODY="Make sure that the update file
+
+;Copy of INSTL constants (Pre-Update check)
+INSTL_PHP_VERSION_NEWER="PHP Version >= %s"
+INSTL_MAGIC_QUOTES_GPC="Magic Quotes GPC Off"
```
No longer needed

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -79,3 +79,44 @@ COM_JOOMLAUPDATE_VIEW_UPLOAD_CAPTIVE_INTRO_BODY="Make sure that the update file
+;Copy of INSTL constants (Pre-Update check)
+INSTL_PHP_VERSION_NEWER="PHP Version >= %s"
+INSTL_MAGIC_QUOTES_GPC="Magic Quotes GPC Off"
+INSTL_REGISTER_GLOBALS="Register Globals Off"
```
no longer needed

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -79,3 +79,44 @@ COM_JOOMLAUPDATE_VIEW_UPLOAD_CAPTIVE_INTRO_BODY="Make sure that the update file
+INSTL_SAFE_MODE="Safe Mode"
+INSTL_DISPLAY_ERRORS="Display Errors"
+INSTL_FILE_UPLOADS="File Uploads"
+INSTL_MAGIC_QUOTES_RUNTIME="Magic Quotes Runtime"
```
no longer needed

brianteeman commented:
### Check for database support.

Is this check enough? The site could be installed on sql but this is not supported in 4.0

roland-d commented:
@brianteeman Very good point about the SQL Server installation. We will check this as well.

SamuelSchepp committed:
Fixed spelling and language keys

brianteeman commented on media/com_joomlaupdate/css/style.css:
```diff
@@ -0,0 +1,10 @@
+    display: inline-block;
+    width: 40%;
+    vertical-align: top;
+    margin-right: 20pt;
```
Not sure if it matters but nowhere else in the core do we use pt as a unit for size - its always px

SamuelSchepp committed:
Changed pt to px in CSS

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1001,309 @@ public function removePackageFiles()
+		$setting->recommended = true;
+		$settings[] = $setting;
+
+		// Check for magic quotes runtimes.
```
is there any point in checking for this. magic quotes runtimes are not present on the supported php version for joomla 4 so if the phpversion check passes then this will always pass - if you agree then dont forget to remove the language string as well

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1001,309 @@ public function removePackageFiles()
+	{
+		$settings = array();
+
+		// Check for safe mode.
```
is there any point in checking for this. safe mode is not present on the supported php version for joomla 4 so if the phpversion check passes then this will always pass - if you agree then dont forget to remove the language string as well

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1001,309 @@ public function removePackageFiles()
 
 	/**
-	 * Get an array with URLs to update servers for a given extension ID
+	 * Get an URL to update servers for a given extension ID
```
Better english would be 
Get the URL to the update server for a given extension ID

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1127,10 +1314,73 @@ private function getUpdateSitesLocations($extension_id)
+	}
+
+	/**
+	 * Method to check 3rd party extensions for compatibility.
```
can we have "non core" instead of 3rd party please.

brianteeman commented:
Sorry I might have jumpe3d the gun a bit with some of my comments. I was assuming that this would only be for updates from 3.9 to 4 but of course it can be for updates from 3.9.0 to 3.9.1 in which case the comments I made about supported php version are not correct. I would suggest that instead of removing those things I highlighted we check the current php version first (i am assuming we cant check which version we are updating to - but might be wrong) and  only check those things if we are on <7.0

SamuelSchepp committed:
Changed database check to only check if current database type is compatible with J! 4.

SamuelSchepp committed:
Re-Added checks, which depend on specific joomla or php versions

SamuelSchepp commented:
We have access to the required PHP version via the update xml file. We now only show specific checks, if they are usefull for the target PHP or Joomla version.

- Database Types if Joomla version is >= 4.
- Magic Quotes (+ Runtime), Register Globals and Safe Mode if the target PHP version is < 7.

brianteeman commented:
@SamuelSchepp great

wilsonge commented on administrator/components/com_joomlaupdate/controllers/update.php:
```diff
@@ -470,4 +470,35 @@ public function finaliseconfirm()
+		}
+		catch (Exception $e)
+		{
+			echo new JResponseJson($e);
```
If creating a `JResponseJson` response here fails is it really worth trying again?

brianteeman commented:
I dont think your php version check is working

<img width="385" alt="screenshotr10-55-44" src="https://user-images.githubusercontent.com/1296369/29815534-da830402-8ca8-11e7-9c46-df5073903651.png">

Quy commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -79,3 +79,46 @@ COM_JOOMLAUPDATE_VIEW_UPLOAD_CAPTIVE_INTRO_BODY="Make sure that the update file
+COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_BREAK="*Extensions marked with <span class='label label-important'>No</span> or <span class='label'>Missing Compatibility Tag</span> might break your website. Please consult with the developer before upgrading."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_MISSING_TAG="*Extensions marked with <span class='label'>Missing Compatibility Tag</span> indicate the developer did not add a compatibility tag in the respective extension's XML."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_UPDATE_REQUIERED="*Extensions marked with <span class='label label-warning'>Yes (X.X.X)</span> might require an update."
```
Spelling `REQUIERED`

Quy commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -79,3 +79,46 @@ COM_JOOMLAUPDATE_VIEW_UPLOAD_CAPTIVE_INTRO_BODY="Make sure that the update file
 COM_JOOMLAUPDATE_VIEW_UPLOAD_PACKAGE_FILE="Joomla package file"
 COM_JOOMLAUPDATE_XML_DESCRIPTION="Updates Joomla to the latest version with one click."
+
+;Pre-Update check
```
Sort keys in alpha order.

wilsonge commented on administrator/components/com_joomlaupdate/controllers/update.php:
```diff
@@ -470,4 +470,35 @@ public function finaliseconfirm()
+		$model = $this->getModel('default');
+		$updateFileUrl = $model->fetchCompatibility($extensionID, $joomlaTargetVersion);
+
+		// ToDo: Change to native function
```
See https://github.com/joomla/joomla-cms/blob/6e07ca6fadd8a5d1db3337cf0357d782646f5411/administrator/components/com_config/controller/application/sendtestmail.php#L28-L30 for how to do it

SamuelSchepp commented:
@brianteeman This is caused by a missing or wrong <php_minimum> tag. We just fixed this in the test instructions. It should say 7.0.

Quy commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -123,3 +128,119 @@
+						<?php echo JText::_($option->state ? 'JYES' : 'JNO'); ?>
+						<?php if ($option->notice) : ?>
+							<span class="icon-info icon-white hasTooltip" title="<?php echo $option->notice; ?>"></span>
+						<?php endif;?>
```
Add space after `;`

brianteeman commented:
@SamuelSchepp thanks - that resolved that issue

SamuelSchepp committed:
Fixed raw header method.
Fixed potential failing JResponseJson().
Fixed some double quotes.
Fixed some phpcs violation.
Sorted language keys alphabetically.
Fixed a language key spelling error.

wilsonge committed:
Small tweak

Quy commented on administrator/components/com_joomlaupdate/views/default/tmpl/default.php:
```diff
@@ -15,6 +15,12 @@
 JHtml::_('script', 'com_joomlaupdate/default.js', array('version' => 'auto', 'relative' => true));
+JHtml::_('stylesheet', 'com_joomlaupdate/style.css', array('version' => 'auto', 'relative' => true));
+
+JText::script("JYES");
```
Use single quote

Quy commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -123,3 +128,119 @@
+	</legend>
+	<table class="table">
+		<tbody>
+		<?php foreach ($this->phpOptions as $option): ?>
```
Add space before `:`. Do the same in other instances.

zero-24 commented on administrator/components/com_joomlaupdate/controllers/update.php:
```diff
@@ -470,4 +470,37 @@ public function finaliseconfirm()
+	 * Prints an JSON string.
+	 * Called from JS.
+	 *
+	 * @return void
```
please add a since tag to this doc block `@since __DEPLOY_VERSION__`

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1001,336 @@ public function removePackageFiles()
 	 *
 	 * @since   __DEPLOY_VERSION__
 	 */
-	public function getExtensions($latest_version)
```
Can we drop a public method?

wilsonge replied:
Yes because it's new from https://github.com/joomla/joomla-cms/pull/16494 so it's never been in a stable version

zero-24 commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -123,3 +128,119 @@
 </fieldset>
+
+<h2>
+	Joomla!
```
Is there a reason to hardcode Joomla?

zero-24 commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php:
```diff
@@ -123,3 +128,119 @@
+				<td class="extension-check"
+					data-extension-id="<?php echo $extension->extension_id; ?>"
+					data-extension-current-version="<?php echo $extension->version; ?>">
+					<img src="../media/system/images/mootree_loader.gif">
```
What about using: https://github.com/joomla/joomla-cms/blob/staging/media/jui/images/ajax-loader.gif and IMO there is also a API to render such loader?

zero-24 commented on media/com_joomlaupdate/js/default.js:
```diff
@@ -11,3 +11,120 @@ function extractionMethodHandler(target, prefix)
 	});
 }
+
+(function($, document, window) {
```
please include that to the minified version too.

zero-24 commented on media/com_joomlaupdate/css/style.css:
```diff
@@ -0,0 +1,10 @@
+fieldset.col {
```
I'm not sure if that is a good idea to include css on a component level. IMO this can be moved to the template with a special indentifier?

SamuelSchepp replied:
We removed css completely. It is now based on boostrap grid.

zero-24 committed:
Apply some major CS changes to this file.

zero-24 commented:
I have just left some coments + applyed some major CS changes to this file: https://github.com/joomla/joomla-cms/pull/17742/commits/30f28f8da2923a12d613815df4eea2a18c3382ab

DrDreave committed:
Remove double quotes

DrDreave committed:
Fix docblock

SamuelSchepp committed:
fixed JApplication

SamuelSchepp committed:
Merge branch '3.9-dev-installation' of https://github.com/SamuelSchepp/joomla-cms into 3.9-dev-installation

* '3.9-dev-installation' of https://github.com/SamuelSchepp/joomla-cms:
  Fix docblock
  Remove double quotes

SamuelSchepp committed:
Added js to minified version, Removed hardcoded "Joomla!".

SamuelSchepp committed:
Added behaviour: If database type is not supported, the update page will show "Your host doesn't support the minimum requirements for Joomla".

SamuelSchepp committed:
Typo

SamuelSchepp committed:
Removed custom css. Now based on bootstrap grid.

SamuelSchepp committed:
Moved JS to existing ScriptDecleration. Changed ScriptDecleration to Heredoc syntax

SamuelSchepp committed:
Fixed: Opening parenthesis of a multi-line function call must be the last content on the line.

Kev1n337 commented:
@brianteeman @wilsonge @zero-24 @Quy 
Thank you for your feedback. We finished improving our pull request and corrected everything you pointed out in your reviews. Could you please test again?

brianteeman commented:
Sorry I should have noticed this before but on a small screen you may not even see the check before doing the update - see screenshot

So why not move all of this pre-upgrade check to a tab and make it the very first tab - then there can be no excuse for someone not seeing it

<img width="913" alt="screenshotr15-12-36" src="https://user-images.githubusercontent.com/1296369/29876878-db9fb1e4-8d95-11e7-8c27-371ae27a38b9.png">

brianteeman commented:
Just to see what happened i set the php version required to 7.5 in the xml and below is the screen that i got

<img width="796" alt="screenshotr15-18-26" src="https://user-images.githubusercontent.com/1296369/29877092-8d4b7c16-8d96-11e7-81b5-939fb639a8ce.png">

To me this text suggests that I should either come back later and try again or download it manually and use the upload and update functionality

The text that i dont meet the minimum requirements is buried in the middle of text which is titled **We can't find a download URL**

brianteeman commented:
### Hathor template

I know as a general rule we are not adding new features to hathor but i think its important for this update screen. In general everything works ok apart from the extension compatibiity

<img width="744" alt="screenshotr15-44-18" src="https://user-images.githubusercontent.com/1296369/29878379-4d5b85c0-8d9a-11e7-9ad2-a69ff33eb7d6.png">

I thought this was because the class for the background was missing from the template and the text was white but that wasnt the case. There is no content there to show if something is compatible or not

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -22,6 +22,12 @@
 {
 	/**
+	 * @var   array  $updateInformation  null
+	 * Holds the update information evaluated in getUpdateInformation.
```
please add the since tag

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1012,377 @@ public function removePackageFiles()
+	 */
+	private function getConfiguredDatabaseType()
+	{
+		return JFactory::getApplication()->get('dbtype');
```
hmm why do you add a method for a simple call to the Application? what about replacing the functon call with that one line?

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -1001,122 +1012,377 @@ public function removePackageFiles()
 	 *
 	 * @since   __DEPLOY_VERSION__
 	 */
-	public function getExtensions($latest_version)
```
my point is still open can we remove a public function?

mbabker replied:
Components are not bound to the same B/C rules as libraries.  So yes, if it is deemed appropriate to do so, stuff may be removed.

wilsonge replied:
Michael is right. But even if that wasn't ok. This function is brand new from #16494 (which was also merged into the 3.9 branch) so it's never been in a stable version

dgrammatiko commented on media/com_joomlaupdate/js/default.js:
```diff
@@ -11,3 +11,120 @@ function extractionMethodHandler(target, prefix)
+     *
+     * @type {Object}
+     */
+    var PreUpdateChecker = {};
```
@SamuelSchepp can you replace this 
```js
 var PreUpdateChecker = {};
```
with 
```js
Joomla.PreUpdateChecker = {};
```

dgrammatiko replied:
@SamuelSchepp Ignore this comment, just realised that these are inside an IIFE

wilsonge commented:
<img width="639" alt="screen shot 2017-08-31 at 00 16 08" src="https://user-images.githubusercontent.com/1986000/29899152-a7d29e8e-8de1-11e7-812c-f77a9fdf6ebd.png">

This looks awesome. I managed to successfully test with extensions that supposedly do and don't support Joomla 4.

I agree with @brianteeman let's add some hathor styling for this case. I do also agree that there's probably some styling and text tweaks to be made, but given you guys finish tomorrow (I believe) let's get this finished and merged so everyone can make pull requests on the finished product :)

Also I'm very impressed you've managed to make @dgt41 happy with your javascript on the first go. It's a hard feat to achieve!

brianteeman commented:
If for any reason you dont manage to complete it then it shouldn't be too hard to add the finishing touches

SamuelSchepp committed:
Fixed issues with hathor template,
Moved Pre-Update check into own Tab (tab only shows if update available),
Added missing @since

Kev1n337 commented:
Thanks again for your suggestions! We fixed the issue with the hathor template and added the missing since-tag.

@brianteeman We also moved the Compatibility Check to a tab and display it in the first place.

We also discussed your suggestion about the **We can't find a download URL** notice when a higher PHP version is required. We agree that this is not the best error description, but we didn't change anything so testing the same thing in staging or the 3.9-dev branch results in the same error description. The only thing we implemented is showing this screen too, when your db-type is not supported by the new update (sqlsrv or sqlazure in J!4) instead of allowing the user to update. More intuitive error descriptions for different situations should be part of a future pull request.

SamuelSchepp committed:
fixed missing language loading in some cases

SamuelSchepp commented:
Could you please test again, as we are (hopefully) done.

SamuelSchepp committed:
Fixed PHP errors on language loading, added support for file & library

brianteeman commented:
Other than the "We can't find a download URL notice when a higher PHP version is required."
 issue which does need to be resolved somehow all the issues I have raised have been resolved.

@wilsonge I guess it can be merged as is and then a new Release Blocker issue created for the above which will need to be resolved.

wilsonge commented:
Thankyou so much to the @icampus students - @SamuelSchepp, @Kev1n337, @DrDreave for all the hard work you've put into this. It's a great effort!

Kev1n337 commented:
I just opened a new pull request handling the new error descriptions, instead of always showing "We can't find a download url".
#17849

nikosdion commented:
If I have a package extension which contains ten extensions and the update URL is only for the package am I correct in understanding that will show all ten extensions as "no compatibility information"? If not, what happens if an extension does not have an update XML info URL at all? Do we simply ignore it on update and hope for the best? I'm sure it's discussed somewhere, I just can't find where. 

Also, I came here from the call to action you posted on Facebook. The information on this PR are not sufficient for a developer to understand what the call to action asks them to do. Please clarify. Do we have to just make sure we have a target platform in our update files or do we need to do something more?

SamuelSchepp commented:
All extensions in a package have a package ID, which points to the root extension, which should have the update XML url. In this case, this is the only entry we show.

Extensions without any update XML url will show "Missing Compatibility Tag". Here we have no information and should consult the developer or "hope for the best".

You can test your Update XML file, by applying the test instructions. You file needs a target platform, version and download url. Just make sure the update checker shows the correct information for your extension.

Some discussion happened here #16494.

nikosdion commented:
> All extensions in a package have a package ID, which points to the root extension

So, if I understand correctly, if they don't (the user extracted the package and installed the extensions manually) they will appear as "Missing Compatibility Tag. OK, I know how to work around that in my code that checks the health of update sites.

> Extensions without any update XML url will show "Missing Compatibility Tag". Here we have no information and should consult the developer or "hope for the best".

Thank you for confirming. I will need to document that for my users.

> You can test your Update XML file, by applying the test instructions. 

The test instructions tell me how to convince Joomla! that there is a version 4 with such and such minimum requirements. I already know how to do that. _I wrote Joomla! Update_ :D The reason I added the custom update source feature was exactly that kind of testing. But that's not what I asked.

I am asking what do I need on *my extension's* XML file. I can infer from the code that I just need to add a target platform and target a specific Joomla! version. Reading #16494 -which I had done before I commented here- I also got no definitive reply to my question. Again, I inferred from the code. Can you please confirm that I just need a target platform in *my extension's* update XML file as accessed by the Update Site URL? Thanks.

SamuelSchepp commented:
You need the tags <code>version</code>, <code>downloadurl</code> and <code>targetplatform</code>. We built upon the logic of JUpdate and these tags are needed for JUpdate to accept an update.

zero-24 commented:
This has been documented: https://docs.joomla.org/Pre-Update_Check by @bayareajenn and now been updated to the new screenshots by me.

bayareajenn commented:
> 
> 
> This has been documented: https://docs.joomla.org/Pre-Update_Check by @bayareajenn and now been updated to the new screenshots by me.

Thank you, Tobias.
